Improve application

horcsinbalint opened the pull request:
## Description
Improved some parts of the application process

## Changes
- Informational texts are now stored in the database and not hard-coded
- Improved experience for users who already have an account
- Allow edit of submission status by sysadmins

horcsinbalint committed:
Added texts that can be configured from the UI

Texts on some places during the application process were edited multiple times last year which was not ideal as they were hard-coded. And we also had an issue if we needed to display information based on which workshop the user applied to.
In this commit I extended the functionality of the `custom` table, renamed it to `configurable_texts` which can be used for storing text for keys and potentially workshops. I added also an admin interface where these settings can be changed.

horcsinbalint committed:
Streamlined application process if someone has an account

Added a button that allows any user that has an account to start the application process.
There has been issues when collegists accidentally started the application process by clicking on a `/application`, added a confirmation for it.

horcsinbalint committed:
Allow submission and unsubmission for sysadmins

Now sysadmins can undo submissions if they are needed (it should not be used but in cases where the student does not receive a Neptun code, it can be useful).
Also made the ApplicationPolicy more understandable.

horcsinbalint committed:
Fix bugs

horcsinbalint commented on app/Http/Controllers/Auth/AdmissionController.php:
```diff
@@ -178,6 +175,22 @@
             $this->storeFile($request, $application->user);
             Mail::bcc($application->committeeMembers())->queue(new ApplicationFileUploaded($request->get('name'), $application));
         }
+        if ($request->has('submit')) {
+            $this->authorize('editSubmissionStatus', Application::class);
+            $application->update(
+                [
+                    "submitted" => 1
+                ]
+            );
+        }
+        if ($request->has('unsubmit')) {
+            $this->authorize('editSubmissionStatus', Application::class);
+            $application->update(
+                [
+                    "submitted" => 0
+                ]
+            );
+        }
```
For human reviewers: `before` is present

horcsinbalint commented on app/Http/Controllers/Auth/AdmissionController.php:
```diff
@@ -210,18 +223,18 @@ public function indexFinalize(): View
             foreach ($admitted as $application) {
                 $application->user->update(['verified' => true]);
-                if($application->admitted_for_resident_status) {
+                if ($application->admitted_for_resident_status) {
```
For human reviewers: we should check on it later

BertalanD replied:
Hm, we should be able to safely delete the print account, right? We are only deleting unverified users, so existing tenants or extern-to-resident reapplicants shouldn't be touched.

And the cascade delete approach is probably simpler than lazily creating the printer accounts.

horcsinbalint replied:
I might misunderstand what's happening but I think `withoutGlobalScope('verified')` only removes the requirement of a user to be verified, and not checks if it should be verified.
So if we had a verified tenant, they would lose their tenant role when starting the application process, and if they aren't admitted, their user would be deleted as they would not have any roles... And they might have a valuable print account.
So removing the tenant role from tenants who want to apply is quite a large problem.

BertalanD replied:
We are also creating internet accounts for new users. Why is that not a problem?

https://github.com/EotvosCollegium/mars/blob/74ca84d193b6e1e1b92c2b15c26d2416af3c5440/app/Models/User.php#L135-L144

horcsinbalint replied:
Internet accesses have cascading delete

horcsinbalint commented on app/Http/Controllers/Auth/ApplicationController.php:
```diff
@@ -42,10 +49,15 @@ public function show(Request $request): View|RedirectResponse
             return redirect()->route('users.tenant-update.show');
         }
 
+        if (user()->isCollegist() &&
+            user()->application()->doesntExist()) {
+            return view("auth.application.confirm_start");
+        }
```
For human reviewers: this is not an issue, there are also other parts of the UI

horcsinbalint commented on resources/views/auth/application/confirm_start.blade.php:
```diff
@@ -0,0 +1,13 @@
+            </form>
+        </div>
+    </div>
+@endsection
```
For human reviewers: this is not an issue, there are also other parts of the UI

horcsinbalint commented on resources/views/auth/application/questions.blade.php:
```diff
@@ -119,18 +119,25 @@
-                    <x-input.textarea id="present"
-                                      text="Amennyiben nem tud jelen lenni a felvételi teljes ideje alatt (kedd-péntek), kérjük itt indoklással jelezze!"
-                                      :value="$user->application->present"  helper="Változás esetén értesítse a titkárságot!"/>
+                    <div class="markdown_with_red">
+                        @markdown(\App\Models\ConfigurableText::getText("APPLICATION_QUESTION2"))
+                        <x-input.textarea id="question_2" :helper="\App\Models\ConfigurableText::getText('APPLICATION_QUESTION2_HELPER')"
+                                          :value="$user->application->question_2"/>
+                    </div>
+                    <div class="markdown_with_red">
+                        @markdown(\App\Models\ConfigurableText::getText("APPLICATION_QUESTION3"))
+                        <x-input.textarea id="question_3" :helper="\App\Models\ConfigurableText::getText('APPLICATION_QUESTION3_HELPER')"
+                                          :value="$user->application->question_3"/>
+                    </div>
+                    <div class="markdown_with_red">
+                        @markdown(\App\Models\ConfigurableText::getText("APPLICATION_QUESTION4"))
+                        <x-input.textarea id="question_4" :helper="\App\Models\ConfigurableText::getText('APPLICATION_QUESTION4_HELPER')"
+                                          :value="$user->application->question_4"/>
+                    </div>
+                    <div class="markdown_with_red">
+                        Amennyiben nem tud jelen lenni a felvételi teljes ideje alatt (kedd-péntek), kérjük itt indoklással jelezze!
+                        <x-input.textarea id="present" :value="$user->application->present"  helper="Változás esetén értesítse a titkárságot!"/>
+                    </div>
```
For human reviewers: an empty text is always created

BertalanD commented on app/Http/Controllers/ConfigurableText/ConfigurableTextController.php:
```diff
@@ -0,0 +1,74 @@
+        'APPLICATION_INFORMATION_PRIOR_TO_FINALIZATION',
+        'APPLICATION_FILES',
+        'APPLICATION_INFORMATION_AFTER_FINALIZATION',
+        'APPLICATION_QUESTION2',
+        'APPLICATION_QUESTION2_HELPER',
+        'APPLICATION_QUESTION3',
+        'APPLICATION_QUESTION3_HELPER',
+        'APPLICATION_QUESTION4',
+        'APPLICATION_QUESTION4_HELPER',
```
Hm, I'm not sure I like this being _partially_ configurable (why is the number of questions fixed?).

I'd prefer if we either keep these hard-coded (I don't think we'd want to change these questions _during_ an ongoing application period) or go for the fully configurable approach (configurable number of questions). This design was strange to begin with (where did question 1 go?!)

BertalanD commented on app/Http/Controllers/Auth/AdmissionController.php:
```diff
@@ -124,7 +124,7 @@ public function index(Request $request)
         $applications = $applications->with('user.educationalInformation')->distinct()->get()->sortBy('user.name');
 
-        if($request->input('return_excel')) {
+        if ($request->input('return_excel')) {
```
argh, unrelated formatting changes

We should run the formatter at once on the whole code base, add it to `.git-blame-ignore-revs`, and then enjoy deepsource-autofix-free PRs

BertalanD commented on resources/views/layouts/navbar.blade.php:
```diff
@@ -33,6 +33,11 @@ class="material-icons">menu</i></a>
         <li class="grey darken-1 white-text" style="padding:5px;line-height:1.5em">@lang('general.debug_descr')</li>
     @endif
 
+    @if(app(\App\Http\Controllers\Auth\ApplicationController::class)->isActive())
+        <li><a class="waves-effect" href="{{ route('application') }}"><i
+                            class="material-icons left">person_search</i>Felvételi jelentkezés</a></li>
+    @endif
```
We need to check this: the existing `UserController::tenantToApplicant` function removes the tenant role (and does some other bookkeeping) for _tenants_ that wish to apply. Is that necessary?

horcsinbalint replied:
Yeah, we should check it, I think that you are correct, I don't think it should be required to remove the tenant role if someone is applying.

horcsinbalint committed:
Remove questions from configurable texts

BertalanD commented on app/Models/ConfigurableText.php:
```diff
@@ -0,0 +1,87 @@
+use Illuminate\Database\Eloquent\Relations\HasMany;
+use Illuminate\Database\Eloquent\Casts\Attribute;
+
+/**
+ * App\Models\Checkout
+ *
+ * @property mixed $key
+ * @property mixed $text
+ * @mixin \Eloquent
+ */
```
the AI is not wrong here

BertalanD commented on database/migrations/2025_07_08_152927_configurable_texts.php:
```diff
@@ -0,0 +1,113 @@
+
+        \DB::table('configurable_texts')->insert([
+            "key" => "APPLICANT_REGISTRATION",
+            "rawtext" => "Itt az Eötvös Collegium felvételijére lehet regisztrálni. Ha vendégként internetet szeretnél igénybe venni, akkor [a vendégregisztrációt kell elvégezned](/register/guest).
```
paging @kamillacziraki

A form más, fix szöveges részein általában magázó formát használunk. Ez policy kérdés, hogy mennyire akarunk lazának/informálisnak tűnni.

kamillacziraki replied:
szerintem akkor itt is magázzunk. valahogy így? 
_____
            "rawtext" => "Itt az Eötvös Collegium felvételijére lehet regisztrálni. Ha vendégként internetet szeretne igénybe venni, akkor [a vendégregisztrációt kell elvégeznie](/register/guest).

kamillacziraki replied:
meg itt a többit is akkor szeintem érdemes lenne, látom nem ez van csak

horcsinbalint committed:
Make text more formal

horcsinbalint committed:
Fix nit caught by AI

horcsinbalint committed:
Fix errors with already existing accounts

horcsinbalint committed:
Improve application page

Renamed the tabs, submission button is only shown if the user has not
submitted the application yet.

kamillacziraki commented:
I really like the more formal texts, as well as the specifications on the necessary documents to upload.

horcsinbalint committed:
Added option for sysadmins to see the profile page of everyone (including applicants)

BertalanD commented on routes/web.php:
```diff
@@ -170,6 +178,11 @@
     Route::post('/network/admin/checkout/income/add', [AdminCheckoutController::class, 'addIncome'])->name('admin.checkout.income.add');
     Route::get('/network/admin/checkout/transaction/delete/{transaction}', [EconomicController::class, 'deleteTransaction'])->name('admin.checkout.transaction.delete');
 
+    Route::prefix('configurable_texts')->name('configurable_texts.')->group(function () {
+        Route::get('/', [ConfigurableTextController::class, 'index'])->name('index');
+        Route::post('/store', [ConfigurableTextController::class, 'store'])->name('store');
+    });
```
No, this is a false positive. See line 87.

horcsinbalint committed:
Fix nits

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 69856ce according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/743
